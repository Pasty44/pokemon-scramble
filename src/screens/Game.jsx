import { useCallback, useEffect, useRef, useState } from 'react';
import { shuffleArray, shuffleString } from 'scripts/util';

const INPUT_COLORS = {
    NEUTRAL: 'black',
    SUCCESS: '#00BB00',
    FAIL: '#DD0000'
};

export default function Game({ pokemonList, reset }) {
    const inputRef = useRef(null);
    const inputColor = useRef(INPUT_COLORS.NEUTRAL);

    // This is so we can clear timer on component unmounting to prevent crashes/memory leaks
    const newPokemonTimeout = useRef(null);
    useEffect(() => {
        return () => {
            if (!!newPokemonTimeout.current) clearTimeout(newPokemonTimeout.current);
        }
    }, []);

    const [currentStreak, setCurrentStreak] = useState(0);
    const [longestStreak, setLongestStreak] = useState(0);

    // Create a shuffled copy of the list prop to iterate through
    const currentPokemonList = useRef(shuffleArray([...pokemonList]));
    const currentIndex = useRef(0);
    const getCurrentPokemonObj = () => ({
        ...currentPokemonList.current[currentIndex.current],
        name: currentPokemonList.current[currentIndex.current].name.toUpperCase(),
        shuffledName: shuffleString(currentPokemonList.current[currentIndex.current].name).toUpperCase()
    });

    const [canType, setCanType] = useState(true);
    const [userGuess, setUserGuess] = useState('');
    const [currentPokemon, setCurrentPokemon] = useState(getCurrentPokemonObj);
    
    const selectNewPokemon = () => {
        // When we reach the end, shuffle and start over
        if (currentIndex.current === currentPokemonList.current.length - 1) {
            currentPokemonList.current = shuffleArray(pokemonList);
            currentIndex.current = 0;
        }
        else {
            currentIndex.current += 1;
        }
        
        setUserGuess('');
        
        inputRef.current.focus();
        
        setCurrentPokemon(getCurrentPokemonObj());
    };

    // Select new pokemon after a delay
    const startNewPokemonTimer = useCallback(() => {
        newPokemonTimeout.current = setTimeout(() => {
            inputColor.current = INPUT_COLORS.NEUTRAL;
            selectNewPokemon();
            setCanType(true);
        }, 1600);
    }, []);
    
    const onType = (updatedGuess, cursorPosition) => {
        if (!canType) return;
        
        setUserGuess(updatedGuess);

        // Prevent weird React behaviour, where putting the cursor in the middle of the guess and
        // typing a character would jump the cursor to the end of the string
        setTimeout(() => {
            inputRef.current.selectionStart = cursorPosition;
            inputRef.current.selectionEnd = cursorPosition;
        }, 0);
        
        if (updatedGuess === currentPokemon.name) {
            inputColor.current = INPUT_COLORS.SUCCESS;
            setCanType(false);
            setCurrentStreak(curr => curr + 1);

            if (currentStreak + 1 > longestStreak) {
                setLongestStreak(curr => curr + 1);
            }
            
            startNewPokemonTimer();
        }
    };

    const skip = useCallback(() => {
        inputColor.current = INPUT_COLORS.FAIL;
        setCanType(false);
        setCurrentStreak(0);
        setUserGuess(currentPokemon.name);

        startNewPokemonTimer();
    }, [currentPokemon, startNewPokemonTimer]);

    return (
        <>
            <div id='spriteContainer'>
                {canType && <div id='spriteMask'>?</div>}
                <img src={currentPokemon.url} height='100%' width='100%' alt='pokemon sprite' />
            </div>

            <div className='my-3'>
                {currentPokemon.shuffledName}
            </div>

            <div id='streakContainer'>
                <div className='streak'>
                    <div className='streakType'>
                        Current streak
                    </div>
                    <div>: {currentStreak}</div>
                </div>
                <div className='streak'>
                    <div className='streakType'>
                        Longest streak
                    </div>
                    <div>: {longestStreak}</div>
                </div>
            </div>

            {currentPokemon.name.indexOf(' ') >-1 && (
                <div id='spaceWarning'>* This name contains a space</div>
            )}

            <input
                id='userGuess'
                className='my-1'
                value={userGuess}
                onChange={(e) => onType(e.target.value.toUpperCase(), e.target.selectionStart)}
                ref={inputRef}
                autoFocus
                autoComplete='off'
                style={{ color: inputColor.current }}
            />

            <button id='skipBtn' className='my-1 noSelect' onClick={skip}>
                I DON'T KNOW
            </button>

            <button id='resetBtn' className='my-1 noSelect' onClick={reset}>
                CHANGE GENS
            </button>
        </>
    );
}
