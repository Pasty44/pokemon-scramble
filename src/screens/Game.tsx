import { useCallback, useEffect, useRef, useState } from "react";
import usePokemonQuiz from "hooks/usePokemonQuiz";
import { TPokemon } from "pokemon/types";

const INPUT_COLORS = {
  NEUTRAL: "#000000",
  SUCCESS: "#00BB00",
  FAIL: "#DD0000",
};

export interface IGameProps {
  pokemonList: Array<TPokemon>;
  reset: () => void;
}

export default function Game({ pokemonList, reset }: IGameProps) {
  const quiz = usePokemonQuiz(pokemonList);
  const inputRef = useRef<any>(null);
  const inputColor = useRef(INPUT_COLORS.NEUTRAL);

  const newPokemonTimeout = useRef<ReturnType<typeof setTimeout> | null>(null);
  // This is so we can clear timer on component unmounting to prevent crashes/memory leaks
  useEffect(() => {
    return () => {
      if (!!newPokemonTimeout.current) clearTimeout(newPokemonTimeout.current);
    };
  }, []);

  const [canType, setCanType] = useState(true);
  const [userGuess, setUserGuess] = useState("");

  // Select new pokemon after a delay
  const startNewPokemonTimer = useCallback(() => {
    const selectNewPokemon = () => {
      setUserGuess("");
      inputRef.current.focus();
      quiz.goNext();
      inputColor.current = INPUT_COLORS.NEUTRAL;
      setCanType(true);
    };

    newPokemonTimeout.current = setTimeout(selectNewPokemon, 2000);
  }, [quiz]);

  const onType = (updatedGuess: string, cursorPosition: number | null) => {
    if (!canType) return;

    setUserGuess(updatedGuess);

    // Prevent weird React behaviour, where putting the cursor in the middle of the guess and
    // typing a character would jump the cursor to the end of the string
    setTimeout(() => {
      inputRef.current.selectionStart = cursorPosition;
      inputRef.current.selectionEnd = cursorPosition;
    }, 0);

    if (updatedGuess === quiz.currentPokemon.name) {
      inputColor.current = INPUT_COLORS.SUCCESS;
      setCanType(false);
      quiz.incrementCurrentStreak();
      startNewPokemonTimer();
    }
  };

  const skip = useCallback(() => {
    inputColor.current = INPUT_COLORS.FAIL;
    setCanType(false);
    quiz.resetCurrentStreak();
    setUserGuess(quiz.currentPokemon.name);
    startNewPokemonTimer();
  }, [quiz, startNewPokemonTimer]);

  return (
    <>
      <div id="spriteContainer">
        {canType && <div id="spriteMask">?</div>}
        <img
          src={quiz.currentPokemon.url}
          height="100%"
          width="100%"
          alt="pokemon sprite"
        />
      </div>

      <div className="my-3">{quiz.currentPokemon.shuffledName}</div>

      <div id="streakContainer">
        <div className="streak">
          <div className="streakType">Current streak</div>
          <div>: {quiz.currentStreak}</div>
        </div>
        <div className="streak">
          <div className="streakType">Longest streak</div>
          <div>: {quiz.longestStreak}</div>
        </div>
      </div>

      {quiz.currentPokemon.name.indexOf(" ") > -1 && (
        <div id="spaceWarning">* This name contains a space</div>
      )}

      <input
        id="userGuess"
        className="my-1"
        value={userGuess}
        onChange={(e) =>
          onType(e.target.value.toUpperCase(), e.target.selectionStart)
        }
        ref={inputRef}
        autoFocus
        autoComplete="off"
        style={{ color: inputColor.current }}
      />

      <button id="skipBtn" className="my-1 noSelect" onClick={skip}>
        I DON'T KNOW
      </button>

      <button id="resetBtn" className="my-1 noSelect" onClick={reset}>
        CHANGE GENS
      </button>
    </>
  );
}
