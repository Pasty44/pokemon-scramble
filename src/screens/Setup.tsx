import { useState } from "react";
import pokemon from "pokemon/all";
import { TPokemon } from "pokemon/types";

export interface ISetupProps {
  startGame: (list: Array<TPokemon>) => void;
}

export default function Setup({ startGame }: ISetupProps) {
  const [gens, setGens] = useState<{ [key: string]: boolean }>(() => {
    return Object.keys(pokemon).reduce((acc, curr, index) => {
      if (index === 0) acc[curr] = true;
      else acc[curr] = false;
      return acc;
    }, {} as { [key: string]: boolean });
  });

  const onClickStart = () => {
    if (Object.keys(gens).every((key) => gens[key] === false)) {
      return alert("You need to select at least 1 generation");
    }

    startGame(
      Object.keys(gens).reduce((acc, curr) => {
        if (!!gens[curr]) {
          return [...acc, ...pokemon[curr]];
        } else return acc;
      }, [] as Array<TPokemon>)
    );
  };

  return (
    <>
      <img
        id="logo"
        src={process.env.PUBLIC_URL + "/logo.png"}
        alt="pokemon scramble logo"
      />

      <div id="instructions" className="my-1">
        Select which generations are included in the game. Then try to work out
        the scrambled pokemon name!
      </div>

      <div id="genCheckboxContainer">
        {Object.keys(pokemon).map((key) => (
          <div
            className="genCheckbox noSelect"
            key={key}
            onClick={() => setGens({ ...gens, [key]: !gens[key] })}
            style={{
              border: `3px solid ${!!gens[key] ? "black" : "grey"}`,
              color: !!gens[key] ? "black" : "grey",
            }}
          >
            GEN {key}
          </div>
        ))}
      </div>

      <button id="startBtn" className="my-1 noSelect" onClick={onClickStart}>
        START GAME
      </button>
    </>
  );
}
