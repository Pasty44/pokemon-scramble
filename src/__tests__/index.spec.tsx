import renderer from "react-test-renderer";
import App from "../";
import Setup from "screens/Setup";
import Game from "screens/Game";

// Mocked so that ReactDOM.render isn't called, which causes issues in jest env
jest.mock("react-dom");

describe("App", () => {
  it("App should render Setup initially", () => {
    const component = renderer.create(<App />);

    const testInstance = component.root;

    expect(testInstance.findByType(Setup)).toBeDefined();

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("App should render Game after start is clicked", () => {
    const component = renderer.create(<App />);

    const testInstance = component.root;

    renderer.act(() => {
      testInstance.findByProps({ id: "startBtn" }).props.onClick();
    });

    expect(testInstance.findByType(Game)).toBeDefined();
  });

  it("App should render Setup after reset is clicked", () => {
    const component = renderer.create(<App />);

    const testInstance = component.root;

    renderer.act(() => {
      testInstance.findByProps({ id: "startBtn" }).props.onClick();
    });

    renderer.act(() => {
      testInstance.findByProps({ id: "resetBtn" }).props.onClick();
    });

    expect(testInstance.findByType(Setup)).toBeDefined();
  });
});
