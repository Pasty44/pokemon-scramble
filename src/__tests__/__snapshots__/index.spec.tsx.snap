// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App should render Setup initially 1`] = `
Array [
  <img
    alt="pokemon scramble logo"
    id="logo"
    src="/logo.png"
  />,
  <div
    className="my-1"
    id="instructions"
  >
    Select which generations are included in the game. Then try to work out the scrambled pokemon name!
  </div>,
  <div
    id="genCheckboxContainer"
  >
    <div
      className="genCheckbox noSelect"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid black",
          "color": "black",
        }
      }
    >
      GEN 
      1
    </div>
    <div
      className="genCheckbox noSelect"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      2
    </div>
    <div
      className="genCheckbox noSelect"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      3
    </div>
    <div
      className="genCheckbox noSelect"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      4
    </div>
    <div
      className="genCheckbox noSelect"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      5
    </div>
    <div
      className="genCheckbox noSelect"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      6
    </div>
    <div
      className="genCheckbox noSelect"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      7
    </div>
    <div
      className="genCheckbox noSelect"
      onClick={[Function]}
      style={
        Object {
          "border": "3px solid grey",
          "color": "grey",
        }
      }
    >
      GEN 
      8
    </div>
  </div>,
  <button
    className="my-1 noSelect"
    id="startBtn"
    onClick={[Function]}
  >
    START GAME
  </button>,
]
`;
